### PR TITLE
[language] promote vector to a builtin type

### PR DESCRIPTION
**Pull request by vgao1996**
## Summary
This promotes `Vector.T` to a builtin type "vector". 

## Test Plan
cargo test

## Progress Tracker
- [x] IR
  - [x] Parser & Compiler
  - [x] E2E Tests
  - [x] Std Modules/Genesis
- [x] Verifier
- [x] `TypeTag` in `libra_types`
- [x] VM
  - [x] Values (Incl. prop testing)
  - [x] Native Functions & Native Structs
  - [x] Serialization/Deserialization
- [x] Source Lang
  - [x] Parser & Compiler
  - [x] E2E Tests
  - [x] Std Modules
- [x] Tools broken by this change
  - [x] bytecode_to_boogie
  - [x] won't fix: cost-synthesis
- [x] Other Broken Tests
- [x] Final Cleanup
- [x] Deploying Breaking Changes
  - [x] Regenerate genesis?
  - [x] Update script hashes?

### CODE REVIEW

File: language/move-lang/src/cfgir/ast.rs [hunk 1]
```diff
@@ -298,22 +298,37 @@ impl UnannotatedExp_ {
...
     pub fn bool(loc: Loc) -> BaseType {
-        Self::builtin(loc, BuiltinTypeName_::Bool)
+        sp(
+            loc,
+            BaseType_::Apply(
+                sp(loc, Kind_::Unrestricted),
+                sp(loc, TypeName_::Builtin(sp(loc, BuiltinTypeName_::Bool))),
+                vec![],
+            ),
+        )
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/ast.rs]: Revert all of these

File: language/move-lang/src/cfgir/ast.rs [hunk 2]
```diff
@@ -298,22 +298,37 @@ impl UnannotatedExp_ {
 }
 
 impl BaseType_ {
-    pub fn builtin(loc: Loc, b_: BuiltinTypeName_) -> BaseType {
```
Comment (tnowacki) [inline language/move-lang/src/cfgir/ast.rs]: Make this take type arguments

File: language/move-lang/src/hlir/ast.rs [hunk 3]
```diff
@@ -234,22 +234,37 @@ pub enum ExpListItem {
...
     pub fn bool(loc: Loc) -> BaseType {
-        Self::builtin(loc, BuiltinTypeName_::Bool)
+        sp(
+            loc,
+            BaseType_::Apply(
+                sp(loc, Kind_::Unrestricted),
+                sp(loc, TypeName_::Builtin(sp(loc, BuiltinTypeName_::Bool))),
+                vec![],
+            ),
+        )
```
Comment (tnowacki) [inline language/move-lang/src/hlir/ast.rs]: Revert

File: language/move-lang/src/cfgir/ast.rs (hunk 2 above)
Comment (tnowacki) [reply language/move-lang/src/cfgir/ast.rs]: Then you can replicate logic around inferring the kind. Case on the kind provided by `b_.kind()`, if its unrestricted. Infer the kind from the type arguments.

File: language/move-lang/src/naming/ast.rs [hunk 4]
```diff
@@ -310,22 +314,26 @@ impl BuiltinTypeName_ {
...
+            Vector => arg_kinds[0],
         }
-    }
+    }*/
+    // ,
 
-    pub fn tparam_constraints(&self, _loc: Loc) -> Vec<Kind> {
+    pub fn tparam_constraints(&self, loc: Loc) -> Vec<Kind> {
         use BuiltinTypeName_::*;
         // Match here to make sure this function is fixed when collections are added
```
Comment (tnowacki) [inline language/move-lang/src/naming/ast.rs]: remove comment

File: language/move-lang/src/naming/ast.rs [hunk 5]
```diff
@@ -376,34 +384,55 @@ impl BuiltinFunction_ {
...
-    pub fn builtin(loc: Loc, b_: BuiltinTypeName_) -> BaseType {
-        let kind = sp(loc, b_.kind());
+    pub fn builtin(loc: Loc, b_: BuiltinTypeName_, ty_args: Vec<BaseType>) -> BaseType {
+        use BuiltinTypeName_::*;
+
+        let kind = match b_ {
+            U8 | U64 | U128 | Address | Bool | Bytearray => Kind_::Unrestricted,
+            // TODO: this can panic.
+            Vector => ty_args[0].value.kind_(),
+        };
```
Comment (tnowacki) [inline language/move-lang/src/naming/ast.rs]: Same as above. BuiltinTypeName should still give a kind.

File: language/move-lang/src/naming/translate.rs [hunk 6]
```diff
@@ -616,10 +616,10 @@ fn base_type(context: &mut Context, ty: E::SingleType) -> N::BaseType {
                 NB::Apply(resource_opt, tn, base_types(context, tys))
             }
             Some(RT::BuiltinType) => {
+                let ty_args = base_types(context, tys);
                 let bn_ = N::BuiltinTypeName_::resolve(&n.value).unwrap();
-                let kind_ = bn_.kind();
-                let tn = sp(loc, NN::Builtin(sp(loc, bn_)));
-                NB::Apply(Some(sp(loc, kind_)), tn, base_types(context, tys))
+                // TODO: get rid of this return?
+                return NB::builtin(loc, bn_, ty_args);
```
Comment (tnowacki) [inline language/move-lang/src/naming/translate.rs]: Remove the return by adding a function `builtin_` in BaseType that doesn't have the loc

File: language/move-lang/src/shared/mod.rs [hunk 7]
```diff
@@ -243,10 +243,10 @@ pub const fn sp<T>(loc: Loc, value: T) -> Spanned<T> {
 /// Macro used to create a tuple-like pattern match for Spanned
 macro_rules! sp {
     (_, $value:pat) => {
-        Spanned { value: $value, .. }
+        Spanned { value: $value, loc: _ }
     };
     ($loc:pat, _) => {
-        Spanned { loc: $loc, .. }
+        Spanned { loc: $loc, value: _ }
```
Comment (tnowacki) [inline language/move-lang/src/shared/mod.rs]: Mega nit. unrelated change to the PR. Its fine here, but try to avoid these if possible :)

File: language/move-lang/src/to_bytecode/translate.rs [hunk 8]
```diff
@@ -642,6 +642,9 @@ fn base_type(context: &mut Context, sp!(_, bt_): G::BaseType) -> Result<F::Signa
 
         B::Apply(_, sp!(_, TN::Builtin(sp!(_, BT::Bool))), _) => ST::Bool,
         B::Apply(_, sp!(_, TN::Builtin(sp!(_, BT::Bytearray))), _) => ST::ByteArray,
+        B::Apply(_, sp!(_, TN::Builtin(sp!(_, BT::Vector))), mut args) => {
+            ST::Vector(Box::new(base_type(context, args.pop().unwrap())?))
```
Comment (tnowacki) [inline language/move-lang/src/to_bytecode/translate.rs]: might want to assert the length of args is 1

File: language/move-lang/src/naming/ast.rs [hunk 9]
```diff
@@ -376,34 +384,55 @@ impl BuiltinFunction_ {
...
 
-    pub fn builtin(loc: Loc, b_: BuiltinTypeName_) -> BaseType {
-        let kind = sp(loc, b_.kind());
+    pub fn builtin(loc: Loc, b_: BuiltinTypeName_, ty_args: Vec<BaseType>) -> BaseType {
+        use BuiltinTypeName_::*;
+
+        let kind = match b_ {
+            U8 | U64 | U128 | Address | Bool | Bytearray => Kind_::Unrestricted,
+            // TODO: this can panic.
+            Vector => ty_args[0].value.kind_(),
```
Comment (tnowacki) [inline language/move-lang/src/naming/ast.rs]: This is your bug. This needs to be `None`

Due to type inference, you cannot inspect the kind at this point. (in any ast past typing, it is fine to do so)

File: language/move-lang/src/shared/mod.rs (hunk 7 above)
Comment (vgao1996) [reply language/move-lang/src/shared/mod.rs]: Ah sorry this was merely a way to workaround a bug in rust-analyzer for myself. Won't include this in the final PR.

File: language/vm/vm-runtime/src/code_cache/module_cache.rs [hunk 10]
```diff
@@ -234,14 +239,14 @@ impl<'alloc> VMModuleCache<'alloc> {
                 Ok(Type::Struct(struct_def))
             }
             SignatureToken::Reference(sub_tok) => {
-                let inner_ty =
+                let inner_layout =
```
Comment (dariorussi) [inline language/vm/vm-runtime/src/code_cache/module_cache.rs]: layout is a horrible name for types, it would be good to revert this change and avoid needless churn

Comment (vgao1996) [reply language/vm/vm-runtime/src/code_cache/module_cache.rs]: So just to clarify: I use the name "layout" because this is what the data structure currently is -- it is **not** a nominal type but a description of the data layout of the values of a type. I would even suggest we rename `enum Type` to `Layout` or `TypeLayout` or `DataLayout` or `ValueLayout` to reduce confusion.

On the other hand, if you ask me what `fn resolve_signature_token` should return I will agree it should a "type". More precisely a data structure that contains both the fully qualified type name & data layouts. (Is this the reason you raised the question?) My understanding is that we will eventually change to this model and rewrite the existing resolution/caching logic drastically. Based on this I don't see much value naming variables in a way that "pretends" these are types (which they are not). Nonetheless I really don't have strong opinions on this, and I believe you are likely more informed than me on what the future VM types will look like, so I'll leave the final call to you.

Comment (dariorussi) [reply language/vm/vm-runtime/src/code_cache/module_cache.rs]: I guess I see your point... 
I suppose I was reacting to the change with the following in mind (albeit not articulated at all):
1- in a PR like this and given the changes we are going to make, having "random" renaming just makes the PR heavier with no benefit
2 - as you mentioned the module cache should return types and nothing related to layout strictly speaking

I would appreciate if you could keep changes to a minimum particularly when there are much more significant things to follow up in this PR.
Clean ups can be done at a later time when not exactly in the context of what we are doing here

Comment (vgao1996) [reply language/vm/vm-runtime/src/code_cache/module_cache.rs]: I get your point. I'll remove some name/style changes and try follow the "local" naming conventions to minimize the mental burden on the reviewers. Thanks for pointing this out.

**Comment by vgao1996** [comment]
@wrwg @shazqadeer is move-prover broken on CI? On my machine all move-prover tests pass. (I installed boogie & z3 with @junkil-park's help.)

**Comment by wrwg** [comment]
> @wrwg @shazqadeer is move-prover broken on CI? On my machine all move-prover tests pass. (I installed boogie & z3 with @junkil-park's help.)

Unless @junkil-park  has changed something recently regards Ci integration, boogie should never be executed in Ci, because the environment variables BOOGIE_EXE and Z3_EXE aren't set. No idea also what happens here w.r.t. MIRAI. 

One other hint: I think Shaz made some change to the logic of calling boogie recently. But this code should not be hit if BOOGIE_EXE/Z3_EXE is not set.

**Comment by vgao1996** [comment]
@wrwg talked to @junkil-park offline and he did enable it recently. We're actually good right now -- the errors were caused by an early build. @junkil-park looked into the latest run and it was successful. 

Thank you both for helping!

**Comment by junkil-park** [comment]
> [move-prover] `cargo test` for Move Prover (`language/move-prover/bytecode-to-boogie`) failed with the following error message:
> 
> ```
> 
> ```

Just in case you'd like know what error caused this comment, you can see the log: https://github.com/mimoo/mirai-bot/runs/471358231?check_suite_focus=true#step:7:541

**Comment by vgao1996** [comment]
@junkil-park it seems to be from an old commit so we should be good.
